### PR TITLE
PT-166788647 FATE efficient maps

### PR DESCRIPTION
**Pull request by UlfNorell**
- [PT-166788647](https://www.pivotaltracker.com/story/show/166788647)
- aeternity/aesophia#121
-  aeternity/aeternity#2646

Adds support for efficient maps in the contract state. This adds a new FATE value `?FATE_STORE_MAP(Cache, Id)` containing a reference to a map saved in the state and a cache of updates to it. When writing the state we go over the terms and allocate store maps for (sufficiently large) maps and updates to old maps. See `aeb_fate_maps:allocate_store_maps/2`. Also some code for doing reference counting to enable garbage collection. The bulk of that code is in the aeternity PR.

Other minor things
- Removes remote tail call instructions
- Adds an arity argument to remote calls. Previously remote calls with wrong arity could mess up the stack, now they fail with a nice error.

### CODE REVIEW

File: src/aeb_fate_maps.erl [hunk 1]
```diff
@@ -124,6 +131,61 @@ write_cache(Key, ?FATE_MAP_TOMBSTONE, Map) ->
...
+%% -- Reference counting -----------------------------------------------------
+
+-type refcount() :: #{id() => pos_integer()}.
+
+-spec refcount_zero() -> refcount().
+refcount_zero() -> #{}.
+
+-spec refcount_diff(refcount(), refcount()) -> refcount().
+refcount_diff(New, Old) ->
+    maps:fold(fun(K, N, C) -> maps:update_with(K, fun(M) -> M - N end, -N, C) end,
```
Comment (lucafavatella) [inline src/aeb_fate_maps.erl]: So this may return a negative refcount? (While the type suggests otherwise.)

File: src/aeb_fate_generate_ops.erl [hunk 2]
```diff
@@ -46,11 +46,11 @@ ops_defs() ->
...
-    , { 'CALL_TR',             16#05,   true,   false,   true,  8, [a,is,a],        call_tr,          {contract, string, integer},     any, "Remote tail call to contract Arg0 and function Arg1 with value Arg2. The types of the arguments has to match the argument typs of the function. And the return type of the called function has to match the type of the current function."}
-    , { 'CALL_GR',             16#06,   true,   false,   true,  8, [a,is,a,a],      call_gr, {contract, string, integer, integer},     any, "Remote call with gas cap in Arg3. Otherwise as CALL_R."}
-    , { 'CALL_GTR',            16#07,   true,   false,   true,  8, [a,is,a,a],     call_gtr, {contract, string, integer, integer},     any, "Remote tail call with gas cap in Arg3. Otherwise as CALL_TR."}
+    , { 'CALL_TR',             16#05,   true,   false,   true,  8, [a,is,a],        call_tr,          {contract, string, integer},     any, "DEPRECATED: Remote tail call to contract Arg0 and function Arg1 with value Arg2. The types of the arguments has to match the argument typs of the function. And the return type of the called function has to match the type of the current function."}
```
Comment (lucafavatella) [inline src/aeb_fate_generate_ops.erl]: Not clear why not removed.

File: src/aeb_fate_generate_ops.erl [hunk 3]
```diff
@@ -46,11 +46,11 @@ ops_defs() ->
...
+    , { 'CALL_GR',             16#06,   true,   false,   true,  8, [a,is,ii,a,a],   call_gr, {contract, string, integer, integer, integer}, any, "Remote call with gas cap in Arg3. Otherwise as CALL_R."}
+    , { 'CALL_GTR',            16#07,   true,   false,   true,  8, [a,is,a,a],     call_gtr, {contract, string, integer, integer},     any, "DEPRECATED: Remote tail call with gas cap in Arg3. Otherwise as CALL_TR."}
```
Comment (lucafavatella) [inline src/aeb_fate_generate_ops.erl]: Not clear why not removed.

File: src/aeb_fate_maps.erl (hunk 1 above)
Comment (UlfNorell) [reply src/aeb_fate_maps.erl]: Yes indeed. The type should really have `integer()` instead of `pos_integer()`.

**Pull request merged**